File: sum/sum.c
/**
 * @file sum.c
 * @author Brian Kirkpatrick (code@tythos.net)
 * @brief See README.md
 * @version 0.1
 * @date 2021-11-06
 * 
 * @copyright Copyright (c) 2021
 * 
 */

extern unsigned char __heap_base;

unsigned int bump_pointer = &__heap_base;

void* malloc(int n) {
    unsigned int r = bump_pointer;
    bump_pointer += n;
    return (void*)r;
}

void free(void* p) {
    // lol
}

int sum(int a[], int len) {
    int sum = 0;
    for (int i = 0; i < len; i++) {
        sum += a[i];
    }
    return sum;
}

File: sum/sum.js
/**
 * @author <code@tythos.net>
 */

require(['../lib/logger'], function(logger) {
  fetch('./sum.wasm')
      .then(response => response.arrayBuffer())
      .then(bytes => WebAssembly.instantiate(bytes, {}))
      .then(instantiation => {
        const exports = instantiation.instance.exports;
        const jsArray = [1, 2, 3, 4];
        const cArrayPointer = exports.malloc(jsArray.length * 4);
        const cArray = new Uint32Array(
            exports.memory.buffer, cArrayPointer, jsArray.length);
        cArray.set(jsArray);
        logger.log(exports.sum(cArrayPointer, cArray.length));
      })
});
